File: listas/python/doubly_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def clear(self):
        self.head = None
        self.tail = None
        self.size = 0

    def is_empty(self):
        return self.head is None

    def get_size(self):
        return self.size

    def add_at_start(self, elem):
        node = Node(elem)
        if self.is_empty():
            self.head = node
            self.tail = node
            self.size += 1
            return
        node.next = self.head
        self.head.prev = node
        self.head = node
        self.size += 1

    def add_at_end(self, elem):
        node = Node(elem)
        if self.is_empty():
            self.head = node
            self.tail = node
            self.size += 1
            return
        node.prev = self.tail
        self.tail.next = node
        self.tail = node
        self.size += 1

    def add_at_position(self, index, elem):
        if index == 0:
            self.add_at_start(elem)
            return
        if index == self.get_size():
            self.add_at_end(elem)
            return
        node = Node(elem)
        current = self.get_node(index - 1)
        node.next = current.next
        node.prev = current
        current.next.prev = node
        current.next = node
        self.size += 1

    def get(self, index):
        current = self.get_node(index)
        return current.value if current else None

    def get_node(self, index):
        if index < 0 or index >= self.get_size():
            return None
        current = self.head
        for _ in range(index):
            current = current.next
        return current

    def contains(self, elem):
        return self.index_of(elem) != -1

    def index_of(self, elem):
        current = self.head
        index = 0
        while current:
            if current.value == elem:
                return index
            current = current.next
            index += 1
        return -1

    def remove_head(self):
        if self.is_empty():
            return None
        node = self.head
        self.head = self.head.next
        if self.head:
            self.head.prev = None
        else:
            self.tail = None
        self.size -= 1
        return node.value

    def remove_tail(self):
        if self.is_empty():
            return None
        node = self.tail
        self.tail = self.tail.prev
        if self.tail:
            self.tail.next = None
        else:
            self.head = None
        self.size -= 1
        return node.value

    def remove_at_position(self, index):
        if self.is_empty() or index < 0 or index >= self.get_size():
            return None
        if index == 0:
            return self.remove_head()
        if index == self.get_size() - 1:
            return self.remove_tail()
        current = self.get_node(index)
        current.prev.next = current.next
        current.next.prev = current.prev
        self.size -= 1
        return current.value

    def remove(self, elem):
        index = self.index_of(elem)
        if index == -1:
            return False
        self.remove_at_position(index)
        return True

    def reverse(self):
        current = self.head
        while current:
            current.next, current.prev = current.prev, current.next
            current = current.prev
        self.head, self.tail = self.tail, self.head

    def to_array(self):
        elements = []
        current = self.head
        while current:
            elements.append(current.value)
            current = current.next
        return elements

File: listas/java/src/DoublyLinkedList.java
public class DoublyLinkedList<T> {
    private NodeDoubly<T> head;
    private NodeDoubly<T> tail;
    private int size;

    public DoublyLinkedList() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public void clear() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public boolean isEmpty() {
        return this.head == null;
    }

    public int getSize() {
        return this.size;
    }

    public void addAtStart(T elem) {
        NodeDoubly<T> node = new NodeDoubly<>(elem);
        if (isEmpty()) {
            this.head = node;
            this.tail = node;
        } else {
            node.setNext(this.head);
            this.head.setPrev(node);
            this.head = node;
        }
        this.size++;
    }

    public void addAtEnd(T elem) {
        NodeDoubly<T> node = new NodeDoubly<>(elem);
        if (isEmpty()) {
            this.head = node;
            this.tail = node;
        } else {
            node.setPrev(this.tail);
            this.tail.setNext(node);
            this.tail = node;
        }
        this.size++;
    }

    public void addAtPosition(int index, T elem) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
        }
        if (index == 0) {
            addAtStart(elem);
            return;
        }
        if (index == size) {
            addAtEnd(elem);
            return;
        }
        NodeDoubly<T> node = new NodeDoubly<>(elem);
        NodeDoubly<T> current = getNode(index - 1);
        node.setNext(current.getNext());
        node.setPrev(current);
        current.getNext().setPrev(node);
        current.setNext(node);
        this.size++;
    }

    public T get(int index) {
        NodeDoubly<T> current = getNode(index);
        if (current != null)
            return current.getValue();
        throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
    }

    private NodeDoubly<T> getNode(int index) {
        if (index < 0 || index >= size)
            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);

        NodeDoubly<T> current = this.head;
        for (int currentIndex = 0; currentIndex < index; currentIndex++) {
            current = current.getNext();
        }
        return current;
    }

    public boolean contains(T elem) {
        return indexOf(elem) != -1;
    }

    public int indexOf(T elem) {
        NodeDoubly<T> current = this.head;
        int index = 0;
        while(current != null) {
            if (current.getValue().equals(elem))
                return index;
            current = current.getNext();
            index++;
        }
        return -1;
    }

    public T removeHead() {
        if (isEmpty())
            throw new IllegalArgumentException("List is empty");

        NodeDoubly<T> node = this.head;
        this.head = this.head.getNext();

        if (this.head == null) {
            this.tail = null;
        } else {
            this.head.setPrev(null);
        }
        this.size--;
        return node.getValue();
    }

    public T removeTail() {
        if (isEmpty())
            throw new IllegalArgumentException("List is empty");

        NodeDoubly<T> node = this.tail;
        this.tail = this.tail.getPrev();

        if (this.tail == null) {
            this.head = null;
        } else {
            this.tail.setNext(null);
        }
        this.size--;
        return node.getValue();
    }

    public T removeAtPosition(int index) {
        if (isEmpty() || index < 0 || index >= size)
            throw new IllegalArgumentException("Invalid index or empty list");

        if (index == 0)
            return removeHead();
        if (index == size - 1)
            return removeTail();

        NodeDoubly<T> current = getNode(index);
        current.getPrev().setNext(current.getNext());
        current.getNext().setPrev(current.getPrev());
        this.size--;
        return current.getValue();
    }

    public boolean remove(T elem) {
        int index = indexOf(elem);
        if (isEmpty() || index == -1) {
            return false;
        }
        removeAtPosition(index);
        return true;
    }

    public void reverse() {
        NodeDoubly<T> temp = null;
        NodeDoubly<T> current = this.head;
        this.tail = this.head;
        while (current != null) {
            temp = current.getPrev();
            current.setPrev(current.getNext());
            current.setNext(temp);
            current = current.getPrev();
        }
        if (temp != null) {
            this.head = temp.getPrev();
        }
    }

    public Object[] toArray() {
        Object[] result = new Object[size];
        NodeDoubly<T> current = this.head;
        int index = 0;
        while (current != null) {
            result[index++] = current.getValue();
            current = current.getNext();
        }
        return result;
    }
}
